File: Source/Core/Entities/CharacterRefill.cs
namespace DesoloZantas.Core.Core.Entities
{
    /// <summary>
    /// Character-themed refill crystals that restore dashes and provide character-specific effects.
    /// Each character (Kirby, Madeline, Badeline, etc.) has their own themed refill variant.
    /// </summary>
    [CustomEntity("Ingeste/CharacterRefill")]
    [Tracked]
    public class CharacterRefill : Entity
    {
        /// <summary>
        /// Character types for themed refills
        /// </summary>
        public enum CharacterType
        {
            Kirby,          // Pink star-shaped, grants float ability
            Madeline,       // Red hair-colored, standard refill
            Badeline,       // Purple themed, grants dream dash charge
            Theo,           // Blue/selfie themed
            Granny,         // Mountain elder themed
            Oshiro,         // Ghost resort themed
            Chara,          // Undertale red soul
            Frisk,          // Undertale yellow soul
            Ralsei,         // Deltarune green soul
            Asriel,         // Rainbow/angel themed
            MetaKnight,     // Dark blue sword themed
            KingDedede,     // Royal hammer themed
            Magolor,        // Lor Starcutter themed
            MageKirby,      // Magic/beam themed
            Custom          // For modders to extend
        }

        // Particle types per character
        public static readonly Dictionary<CharacterType, ParticleType> P_Shatter = new();
        public static readonly Dictionary<CharacterType, ParticleType> P_Regen = new();
        public static readonly Dictionary<CharacterType, ParticleType> P_Glow = new();

        // Components
        private Sprite sprite;
        private Sprite flash;
        private Sprite characterIcon; // Small character icon overlay
        private Image outline;
        private Wiggler wiggler;
        private BloomPoint bloom;
        private VertexLight light;
        private SineWave sine;

        // State
        private Level level;
        private float respawnTimer;
        private bool collected;

        // Properties
        public CharacterType Character { get; private set; }
        public int DashCount { get; private set; }
        public bool OneUse { get; private set; }
        public bool GrantsSpecialAbility { get; private set; }
        public string CustomSpritePath { get; private set; }
        public string CustomSoundEvent { get; private set; }
        public bool CharacterModeOnly { get; private set; }
        public bool RefillStamina { get; private set; }

        static CharacterRefill()
        {
            InitializeParticles();
        }

        private static void InitializeParticles()
        {
            // Kirby - Pink/magenta particles
            CreateParticlesForCharacter(CharacterType.Kirby, 
                Color.HotPink, Color.LightPink, Color.MediumVioletRed);

            // Madeline - Red/auburn particles
            CreateParticlesForCharacter(CharacterType.Madeline,
                new Color(230, 100, 100), new Color(255, 150, 150), Color.Salmon);

            // Badeline - Purple/violet particles
            CreateParticlesForCharacter(CharacterType.Badeline,
                new Color(139, 69, 255), new Color(186, 140, 255), Color.BlueViolet);

            // Theo - Blue particles
            CreateParticlesForCharacter(CharacterType.Theo,
                Color.DodgerBlue, Color.LightSkyBlue, Color.DeepSkyBlue);

            // Granny - Silver/grey particles
            CreateParticlesForCharacter(CharacterType.Granny,
                Color.Silver, Color.LightGray, Color.DarkGray);

            // Oshiro - Ghostly green/teal particles
            CreateParticlesForCharacter(CharacterType.Oshiro,
                new Color(100, 200, 180), Color.Cyan, Color.Teal);

            // Chara - Red soul particles
            CreateParticlesForCharacter(CharacterType.Chara,
                Color.Red, Color.DarkRed, new Color(200, 50, 50));

            // Frisk - Yellow soul particles
            CreateParticlesForCharacter(CharacterType.Frisk,
                Color.Gold, Color.Yellow, Color.LightGoldenrodYellow);

            // Ralsei - Green soul particles
            CreateParticlesForCharacter(CharacterType.Ralsei,
                Color.LimeGreen, Color.LightGreen, Color.ForestGreen);

            // Asriel - Rainbow/white particles
            CreateParticlesForCharacter(CharacterType.Asriel,
                Color.White, Color.Gold, Color.LightCyan);

            // Meta Knight - Dark blue particles
            CreateParticlesForCharacter(CharacterType.MetaKnight,
                Color.DarkBlue, Color.MidnightBlue, new Color(30, 30, 120));

            // King Dedede - Royal blue/gold particles
            CreateParticlesForCharacter(CharacterType.KingDedede,
                Color.RoyalBlue, Color.Gold, Color.Navy);

            // Magolor - Purple/orange particles
            CreateParticlesForCharacter(CharacterType.Magolor,
                new Color(100, 50, 150), Color.Orange, Color.Purple);

            // Mage Kirby - Cyan/magic particles
            CreateParticlesForCharacter(CharacterType.MageKirby,
                Color.Cyan, Color.Aquamarine, Color.DarkCyan);
        }

        private static void CreateParticlesForCharacter(CharacterType character, 
            Color primary, Color secondary, Color tertiary)
        {
            P_Shatter[character] = new ParticleType
            {
                Source = GFX.Game["particles/triangle"],
                Color = primary,
                Color2 = secondary,
                ColorMode = ParticleType.ColorModes.Blink,
                FadeMode = ParticleType.FadeModes.Late,
                LifeMin = 0.5f,
                LifeMax = 1f,
                Size = 1f,
                SpeedMin = 100f,
                SpeedMax = 200f,
                Direction = -(float)Math.PI / 2f,
                DirectionRange = (float)Math.PI
            };

            P_Regen[character] = new ParticleType
            {
                Source = GFX.Game["particles/blob"],
                Color = primary,
                Color2 = tertiary,
                ColorMode = ParticleType.ColorModes.Choose,
                FadeMode = ParticleType.FadeModes.Late,
                LifeMin = 0.5f,
                LifeMax = 1f,
                Size = 0.7f,
                SpeedMin = 10f,
                SpeedMax = 20f,
                SpeedMultiplier = 0.1f,
                Direction = -(float)Math.PI / 2f,
                DirectionRange = (float)Math.PI / 4f
            };

            P_Glow[character] = new ParticleType
            {
                Source = GFX.Game["particles/blob"],
                Color = primary,
                Color2 = Color.White,
                ColorMode = ParticleType.ColorModes.Blink,
                FadeMode = ParticleType.FadeModes.InAndOut,
                LifeMin = 0.6f,
                LifeMax = 1f,
                Size = 0.5f,
                SpeedMin = 4f,
                SpeedMax = 8f,
                DirectionRange = (float)Math.PI * 2f
            };
        }

        public CharacterRefill(Vector2 position, CharacterType character, int dashCount = 1, bool oneUse = false) 
            : base(position)
        {
            Character = character;
            DashCount = Math.Max(1, dashCount);
            OneUse = oneUse;
            GrantsSpecialAbility = GetDefaultSpecialAbility(character);
            RefillStamina = true;

            Initialize();
        }

        public CharacterRefill(EntityData data, Vector2 offset)
            : base(data.Position + offset)
        {
            Character = (CharacterType)data.Int("characterType", 0);
            DashCount = Math.Max(1, data.Int("dashCount", 1));
            OneUse = data.Bool("oneUse", false);
            GrantsSpecialAbility = data.Bool("grantsSpecialAbility", GetDefaultSpecialAbility(Character));
            CustomSpritePath = data.Attr("customSpritePath", "");
            CustomSoundEvent = data.Attr("customSoundEvent", "");
            CharacterModeOnly = data.Bool("characterModeOnly", false);
            RefillStamina = data.Bool("refillStamina", true);

            Initialize();
        }

        private bool GetDefaultSpecialAbility(CharacterType character)
        {
            return character switch
            {
                CharacterType.Kirby => true,      // Float ability
                CharacterType.Badeline => true,   // Dream dash
                CharacterType.Asriel => true,     // Angel wings
                CharacterType.MetaKnight => true, // Sword dash
                _ => false
            };
        }

        private void Initialize()
        {
            Collider = new Hitbox(16f, 16f, -8f, -8f);

            // Get sprite path based on character
            string spritePath = GetSpritePath();
            string spriteId = GetSpriteId();

            // Main sprite
            try
            {
                sprite = GFX.SpriteBank.Create(spriteId);
            }
            catch
            {
                // Fallback to default refill
                sprite = GFX.SpriteBank.Create("refill");
            }
            sprite.Play("idle");
            Add(sprite);

            // Flash effect
            Add(flash = new Sprite(GFX.Game, spritePath));
            flash.Add("flash", "flash", 0.05f);
            flash.OnFinish = delegate { flash.Visible = false; };
            flash.CenterOrigin();
            flash.Visible = false;

            // Wiggler for collection effect
            Add(wiggler = Wiggler.Create(1f, 4f, v =>
            {
                sprite.Scale = Vector2.One * (1f + v * 0.2f);
            }));

            Add(new MirrorReflection());
            
            // Lighting based on character color
            Color lightColor = GetCharacterColor();
            Add(bloom = new BloomPoint(0.8f, 16f));
            Add(light = new VertexLight(lightColor, 1f, 16, 48));
            
            Add(sine = new SineWave(0.6f, 0f));
            sine.Randomize();

            UpdateY();
            Depth = -100;

            // Player collision
            Add(new PlayerCollider(OnPlayer));
        }

        private string GetSpritePath()
        {
            if (!string.IsNullOrEmpty(CustomSpritePath))
                return CustomSpritePath;

            return Character switch
            {
                CharacterType.Kirby => "objects/characterRefill/kirby/",
                CharacterType.Madeline => "objects/characterRefill/madeline/",
                CharacterType.Badeline => "objects/characterRefill/badeline/",
                CharacterType.Theo => "objects/characterRefill/theo/",
                CharacterType.Granny => "objects/characterRefill/granny/",
                CharacterType.Oshiro => "objects/characterRefill/oshiro/",
                CharacterType.Chara => "objects/characterRefill/chara/",
                CharacterType.Frisk => "objects/characterRefill/frisk/",
                CharacterType.Ralsei => "objects/characterRefill/ralsei/",
                CharacterType.Asriel => "objects/characterRefill/asriel/",
                CharacterType.MetaKnight => "objects/characterRefill/metaknight/",
                CharacterType.KingDedede => "objects/characterRefill/dedede/",
                CharacterType.Magolor => "objects/characterRefill/magolor/",
                CharacterType.MageKirby => "objects/characterRefill/magekirby/",
                _ => "objects/refill/"
            };
        }

        private string GetSpriteId()
        {
            return Character switch
            {
                CharacterType.Kirby => "characterRefillKirby",
                CharacterType.Madeline => "characterRefillMadeline",
                CharacterType.Badeline => "characterRefillBadeline",
                CharacterType.Theo => "characterRefillTheo",
                CharacterType.Granny => "characterRefillGranny",
                CharacterType.Oshiro => "characterRefillOshiro",
                CharacterType.Chara => "characterRefillChara",
                CharacterType.Frisk => "characterRefillFrisk",
                CharacterType.Ralsei => "characterRefillRalsei",
                CharacterType.Asriel => "characterRefillAsriel",
                CharacterType.MetaKnight => "characterRefillMetaKnight",
                CharacterType.KingDedede => "characterRefillDedede",
                CharacterType.Magolor => "characterRefillMagolor",
                CharacterType.MageKirby => "characterRefillMageKirby",
                _ => DashCount >= 2 ? "refillTwo" : "refill"
            };
        }

        private Color GetCharacterColor()
        {
            return Character switch
            {
                CharacterType.Kirby => Color.HotPink,
                CharacterType.Madeline => new Color(230, 100, 100),
                CharacterType.Badeline => new Color(139, 69, 255),
                CharacterType.Theo => Color.DodgerBlue,
                CharacterType.Granny => Color.Silver,
                CharacterType.Oshiro => new Color(100, 200, 180),
                CharacterType.Chara => Color.Red,
                CharacterType.Frisk => Color.Gold,
                CharacterType.Ralsei => Color.LimeGreen,
                CharacterType.Asriel => Color.White,
                CharacterType.MetaKnight => Color.DarkBlue,
                CharacterType.KingDedede => Color.RoyalBlue,
                CharacterType.Magolor => new Color(100, 50, 150),
                CharacterType.MageKirby => Color.Cyan,
                _ => Color.White
            };
        }

        private string GetTouchSoundEvent()
        {
            if (!string.IsNullOrEmpty(CustomSoundEvent))
                return CustomSoundEvent;

            return Character switch
            {
                CharacterType.Kirby => "event:/char/kirby/refill_touch",
                CharacterType.Badeline => "event:/game/05_mirror_temple/redbooster_dash",
                CharacterType.Chara => "event:/game/09_core/death_bump",
                CharacterType.Asriel => "event:/final_content/game/19_TheEnd/gigadiamond_touch",
                CharacterType.MetaKnight => "event:/game/general/sword_touch",
                _ => DashCount >= 2 ? "event:/game/general/diamond_touch" : "event:/game/general/diamond_touch"
            };
        }

        public override void Added(Scene scene)
        {
            base.Added(scene);
            level = SceneAs<Level>();
        }

        public override void Update()
        {
            base.Update();

            // Check character mode restriction
            if (CharacterModeOnly)
            {
                string flagName = GetCharacterModeFlag();
                bool isCharacterMode = level?.Session.GetFlag(flagName) ?? false;
                Collidable = isCharacterMode && respawnTimer <= 0f;
                Visible = isCharacterMode;
            }
            else
            {
                Collidable = respawnTimer <= 0f;
            }

            if (respawnTimer > 0f)
            {
                respawnTimer -= Engine.DeltaTime;
                if (respawnTimer <= 0f)
                {
                    Respawn();
                }
            }
            else if (Scene.OnInterval(0.1f) && sprite.Visible)
            {
                var particleType = P_Glow.TryGetValue(Character, out var pt) ? pt : global::Celeste.Refill.P_Glow;
                level?.ParticlesFG.Emit(particleType, 1, Position, Vector2.One * 5f);
            }

            UpdateY();
            light.Alpha = Calc.Approach(light.Alpha, sprite.Visible ? 1f : 0f, 4f * Engine.DeltaTime);
            bloom.Alpha = light.Alpha * 0.8f;

            if (Scene.OnInterval(2f) && sprite.Visible)
            {
                flash.Play("flash", restart: true);
                flash.Visible = true;
            }
        }

        private string GetCharacterModeFlag()
        {
            return Character switch
            {
                CharacterType.Kirby => "kirby_mode",
                CharacterType.Badeline => "badeline_mode",
                CharacterType.Chara => "chara_mode",
                CharacterType.Ralsei => "ralsei_mode",
                CharacterType.Asriel => "asriel_mode",
                _ => $"{Character.ToString().ToLower()}_mode"
            };
        }

        private void Respawn()
        {
            if (!Collidable)
            {
                Collidable = true;
                sprite.Visible = true;
                wiggler.Start();

                var particleType = P_Regen.TryGetValue(Character, out var pt) ? pt : global::Celeste.Refill.P_Regen;
                Audio.Play("event:/game/general/diamond_return", Position);
                level?.ParticlesFG.Emit(particleType, 16, Position, Vector2.One * 2f);
            }
        }

        private void UpdateY()
        {
            float yOffset = sine.Value * 2f;
            sprite.Y = bloom.Y = yOffset;
            flash.Y = yOffset;
        }

        public override void Render()
        {
            if (sprite.Visible)
            {
                sprite.DrawOutline();
            }
            base.Render();
        }

        private void OnPlayer(global::Celeste.Player player)
        {
            if (collected) return;

            // Grant dashes
            player.RefillDash();
            for (int i = 1; i < DashCount; i++)
            {
                if (player.Dashes < player.MaxDashes + DashCount - 1)
                    player.Dashes++;
            }

            // Refill stamina if enabled
            if (RefillStamina)
            {
                player.RefillStamina();
            }

            // Grant special abilities based on character
            if (GrantsSpecialAbility)
            {
                ApplySpecialAbility();
            }

            // Play character-specific sound
            Audio.Play(GetTouchSoundEvent(), Position);
            Input.Rumble(RumbleStrength.Medium, RumbleLength.Medium);

            collected = true;
            Collidable = false;

            Add(new Coroutine(RefillRoutine(player)));
        }

        private void ApplySpecialAbility()
        {
            switch (Character)
            {
                case CharacterType.Kirby:
                    level?.Session.SetFlag("has_float_ability", true);
                    level?.Session.SetFlag("kirby_refill_boost", true);
                    break;
                    
                case CharacterType.Badeline:
                    level?.Session.SetFlag("dream_dash_charged", true);
                    break;
                    
                case CharacterType.Asriel:
                    level?.Session.SetFlag("angel_wings_active", true);
                    break;
                    
                case CharacterType.MetaKnight:
                    level?.Session.SetFlag("sword_dash_ready", true);
                    break;
                    
                case CharacterType.Chara:
                    level?.Session.SetFlag("determination_boost", true);
                    break;
                    
                case CharacterType.Ralsei:
                    level?.Session.SetFlag("heal_aura_active", true);
                    break;
            }
        }

        private IEnumerator RefillRoutine(global::Celeste.Player player)
        {
            Celeste.Celeste.Freeze(0.05f);
            yield return null;

            level?.Shake();
            sprite.Visible = false;
            Depth = 8999;

            // Emit character-specific particles
            var particleType = P_Shatter.TryGetValue(Character, out var pt) ? pt : global::Celeste.Refill.P_Shatter;
            level?.ParticlesFG.Emit(particleType, 5, Position, Vector2.One * 4f, -(float)Math.PI / 2f);
            level?.ParticlesFG.Emit(particleType, 5, Position, Vector2.One * 4f, (float)Math.PI / 2f);

            SlashFx.Burst(Position, 0f);

            if (OneUse)
            {
                RemoveSelf();
            }
            else
            {
                respawnTimer = 2.5f;
                collected = false;
            }
        }
    }
}

File: Loenn/entities/character_refill.lua
local characterRefill = {}

characterRefill.name = "Ingeste/CharacterRefill"
characterRefill.depth = -100

-- Character type options for the dropdown
characterRefill.fieldInformation = {
    characterType = {
        options = {
            ["Kirby"] = 0,
            ["Madeline"] = 1,
            ["Badeline"] = 2,
            ["Theo"] = 3,
            ["Granny"] = 4,
            ["Oshiro"] = 5,
            ["Chara"] = 6,
            ["Frisk"] = 7,
            ["Ralsei"] = 8,
            ["Asriel"] = 9,
            ["Meta Knight"] = 10,
            ["King Dedede"] = 11,
            ["Magolor"] = 12,
            ["Mage Kirby"] = 13,
            ["Custom"] = 14
        },
        editable = false
    },
    dashCount = {
        fieldType = "integer",
        minimumValue = 1,
        maximumValue = 10
    }
}

characterRefill.fieldOrder = {
    "x", "y",
    "characterType",
    "dashCount",
    "oneUse",
    "grantsSpecialAbility",
    "refillStamina",
    "characterModeOnly",
    "customSpritePath",
    "customSoundEvent"
}

-- Multiple placements for quick character selection
characterRefill.placements = {
    {
        name = "kirby",
        alternativeName = "kirby_refill",
        data = {
            characterType = 0,
            dashCount = 1,
            oneUse = false,
            grantsSpecialAbility = true,
            refillStamina = true,
            characterModeOnly = false,
            customSpritePath = "",
            customSoundEvent = ""
        }
    },
    {
        name = "madeline",
        alternativeName = "madeline_refill",
        data = {
            characterType = 1,
            dashCount = 1,
            oneUse = false,
            grantsSpecialAbility = false,
            refillStamina = true,
            characterModeOnly = false,
            customSpritePath = "",
            customSoundEvent = ""
        }
    },
    {
        name = "badeline",
        alternativeName = "badeline_refill",
        data = {
            characterType = 2,
            dashCount = 2,
            oneUse = false,
            grantsSpecialAbility = true,
            refillStamina = true,
            characterModeOnly = false,
            customSpritePath = "",
            customSoundEvent = ""
        }
    },
    {
        name = "theo",
        alternativeName = "theo_refill",
        data = {
            characterType = 3,
            dashCount = 1,
            oneUse = false,
            grantsSpecialAbility = false,
            refillStamina = true,
            characterModeOnly = false,
            customSpritePath = "",
            customSoundEvent = ""
        }
    },
    {
        name = "granny",
        alternativeName = "granny_refill",
        data = {
            characterType = 4,
            dashCount = 1,
            oneUse = false,
            grantsSpecialAbility = false,
            refillStamina = true,
            characterModeOnly = false,
            customSpritePath = "",
            customSoundEvent = ""
        }
    },
    {
        name = "oshiro",
        alternativeName = "oshiro_refill",
        data = {
            characterType = 5,
            dashCount = 1,
            oneUse = false,
            grantsSpecialAbility = false,
            refillStamina = true,
            characterModeOnly = false,
            customSpritePath = "",
            customSoundEvent = ""
        }
    },
    {
        name = "chara",
        alternativeName = "chara_refill",
        data = {
            characterType = 6,
            dashCount = 1,
            oneUse = false,
            grantsSpecialAbility = true,
            refillStamina = true,
            characterModeOnly = false,
            customSpritePath = "",
            customSoundEvent = ""
        }
    },
    {
        name = "frisk",
        alternativeName = "frisk_refill",
        data = {
            characterType = 7,
            dashCount = 1,
            oneUse = false,
            grantsSpecialAbility = false,
            refillStamina = true,
            characterModeOnly = false,
            customSpritePath = "",
            customSoundEvent = ""
        }
    },
    {
        name = "ralsei",
        alternativeName = "ralsei_refill",
        data = {
            characterType = 8,
            dashCount = 1,
            oneUse = false,
            grantsSpecialAbility = true,
            refillStamina = true,
            characterModeOnly = false,
            customSpritePath = "",
            customSoundEvent = ""
        }
    },
    {
        name = "asriel",
        alternativeName = "asriel_refill",
        data = {
            characterType = 9,
            dashCount = 3,
            oneUse = false,
            grantsSpecialAbility = true,
            refillStamina = true,
            characterModeOnly = false,
            customSpritePath = "",
            customSoundEvent = ""
        }
    },
    {
        name = "meta_knight",
        alternativeName = "meta_knight_refill",
        data = {
            characterType = 10,
            dashCount = 2,
            oneUse = false,
            grantsSpecialAbility = true,
            refillStamina = true,
            characterModeOnly = false,
            customSpritePath = "",
            customSoundEvent = ""
        }
    },
    {
        name = "king_dedede",
        alternativeName = "dedede_refill",
        data = {
            characterType = 11,
            dashCount = 2,
            oneUse = false,
            grantsSpecialAbility = false,
            refillStamina = true,
            characterModeOnly = false,
            customSpritePath = "",
            customSoundEvent = ""
        }
    },
    {
        name = "magolor",
        alternativeName = "magolor_refill",
        data = {
            characterType = 12,
            dashCount = 2,
            oneUse = false,
            grantsSpecialAbility = false,
            refillStamina = true,
            characterModeOnly = false,
            customSpritePath = "",
            customSoundEvent = ""
        }
    },
    {
        name = "mage_kirby",
        alternativeName = "mage_kirby_refill",
        data = {
            characterType = 13,
            dashCount = 2,
            oneUse = false,
            grantsSpecialAbility = true,
            refillStamina = true,
            characterModeOnly = false,
            customSpritePath = "",
            customSoundEvent = ""
        }
    }
}

-- Get the texture based on character type
function characterRefill.texture(room, entity)
    local characterType = entity.characterType or 0
    
    -- Map character types to their texture paths
    local textures = {
        [0]  = "objects/characterRefill/kirby/idle00",      -- Kirby
        [1]  = "objects/characterRefill/madeline/idle00",   -- Madeline
        [2]  = "objects/characterRefill/badeline/idle00",   -- Badeline
        [3]  = "objects/characterRefill/theo/idle00",       -- Theo
        [4]  = "objects/characterRefill/granny/idle00",     -- Granny
        [5]  = "objects/characterRefill/oshiro/idle00",     -- Oshiro
        [6]  = "objects/characterRefill/chara/idle00",      -- Chara
        [7]  = "objects/characterRefill/frisk/idle00",      -- Frisk
        [8]  = "objects/characterRefill/ralsei/idle00",     -- Ralsei
        [9]  = "objects/characterRefill/asriel/idle00",     -- Asriel
        [10] = "objects/characterRefill/metaknight/idle00", -- Meta Knight
        [11] = "objects/characterRefill/dedede/idle00",     -- King Dedede
        [12] = "objects/characterRefill/magolor/idle00",    -- Magolor
        [13] = "objects/characterRefill/magekirby/idle00",  -- Mage Kirby
        [14] = "objects/refill/idle00"                      -- Custom/fallback
    }
    
    -- Fallback textures based on dash count
    local fallbackTextures = {
        [0]  = "objects/refill/idle00",       -- 1 dash
        [1]  = "objects/refill/idle00",
        [2]  = "objects/refillTwo/idle00",    -- 2 dash
        [3]  = "objects/refillTwo/idle00",
        [4]  = "objects/refillTwo/idle00",
        [5]  = "objects/refillTwo/idle00",
        [6]  = "objects/refill/idle00",
        [7]  = "objects/refill/idle00",
        [8]  = "objects/refill/idle00",
        [9]  = "objects/refillTwo/idle00",
        [10] = "objects/refillTwo/idle00",
        [11] = "objects/refillTwo/idle00",
        [12] = "objects/refillTwo/idle00",
        [13] = "objects/refillTwo/idle00",
        [14] = "objects/refill/idle00"
    }
    
    local texture = textures[characterType] or textures[14]
    
    -- Try to load character-specific texture, fall back if not found
    -- In Loenn, we'll use the fallback for now until sprites are created
    return fallbackTextures[characterType] or "objects/refill/idle00"
end

-- Get color tint based on character for visual distinction in editor
function characterRefill.color(room, entity)
    local characterType = entity.characterType or 0
    
    local colors = {
        [0]  = {1.0, 0.4, 0.7, 1.0},   -- Kirby - Pink
        [1]  = {0.9, 0.4, 0.4, 1.0},   -- Madeline - Red
        [2]  = {0.55, 0.27, 1.0, 1.0}, -- Badeline - Purple
        [3]  = {0.12, 0.56, 1.0, 1.0}, -- Theo - Blue
        [4]  = {0.75, 0.75, 0.75, 1.0},-- Granny - Silver
        [5]  = {0.4, 0.8, 0.7, 1.0},   -- Oshiro - Teal
        [6]  = {1.0, 0.0, 0.0, 1.0},   -- Chara - Red
        [7]  = {1.0, 0.84, 0.0, 1.0},  -- Frisk - Gold
        [8]  = {0.2, 0.8, 0.2, 1.0},   -- Ralsei - Green
        [9]  = {1.0, 1.0, 1.0, 1.0},   -- Asriel - White
        [10] = {0.0, 0.0, 0.55, 1.0},  -- Meta Knight - Dark Blue
        [11] = {0.25, 0.41, 0.88, 1.0},-- King Dedede - Royal Blue
        [12] = {0.4, 0.2, 0.6, 1.0},   -- Magolor - Purple
        [13] = {0.0, 1.0, 1.0, 1.0},   -- Mage Kirby - Cyan
        [14] = {1.0, 1.0, 1.0, 1.0}    -- Custom - White
    }
    
    return colors[characterType] or colors[14]
end

return characterRefill
